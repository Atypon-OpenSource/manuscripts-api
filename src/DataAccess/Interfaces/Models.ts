/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Invitation,
  Project,
  ContainerInvitation,
  Collaboration,
  UserProfile,
  ContainerRequest,
  Submission,
  ManuscriptNote
} from '@manuscripts/manuscripts-json-schema'

export type ProjectLike = Pick<
  Project,
  Exclude<keyof Project, 'createdAt' | 'updatedAt'>
>

export type InvitationLike = Pick<
  Invitation,
  Exclude<keyof Invitation, 'createdAt' | 'updatedAt'>
>

export type UserProfileLike = Pick<
  UserProfile,
  Exclude<keyof UserProfile, 'createdAt' | 'updatedAt'>
>

export type ContainerInvitationLike = Pick<
  ContainerInvitation,
  Exclude<keyof ContainerInvitation, 'createdAt' | 'updatedAt'>
>

export type CollaborationLike = Pick<
  Collaboration,
  Exclude<keyof Collaboration, 'createdAt' | 'updatedAt'>
>

export type ContainerRequestLike = Pick<
  ContainerRequest,
  Exclude<keyof ContainerRequest, 'createdAt' | 'updatedAt'>
>

export type SubmissionLike = Pick<
  Submission,
  Exclude<keyof Submission, 'createdAt' | 'updatedAt'>
>

export type ManuscriptNoteLike = Pick<
  ManuscriptNote,
  Exclude<keyof ManuscriptNote, 'createdAt' | 'updatedAt'>
>
