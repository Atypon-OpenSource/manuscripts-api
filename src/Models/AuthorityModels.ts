/*!
 * Â© 2024 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Prisma } from '@prisma/client'
import { Step } from 'prosemirror-transform'

export type ReceiveSteps = {
  steps: Prisma.JsonObject[]
  clientID: number
  version: number
}

export type History = {
  steps: Step[] | Prisma.JsonValue[]
  clientIDs: number[]
  version: number
  doc?: Prisma.JsonValue
}

export type DocumentHistory = History & { doc: Prisma.JsonValue | undefined }
export interface ModifiedStep extends Prisma.JsonObject {
  clientID: string
}
