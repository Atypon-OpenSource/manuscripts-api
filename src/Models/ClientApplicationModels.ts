/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createHash } from 'crypto'
import { pem2jwk } from 'pem-jwk'

import { ScopedAccessTokenConfiguration } from '../Config/ConfigurationTypes'
import { ValidationError } from '../Errors'

/**
 * Represents all acceptable fields for client application objects (client application = manuscripts-api client).
 */
export interface ClientApplication {
  /**
   * A unique application id.
   */
  _id?: string

  _type: string

  /**
   * A secret for application.
   */
  secret: string | null

  /**
   * The name of the application.
   */
  name: string

  /**
   * Details of application.
   */
  details?: string
}

export function scopeConfigurationsFromSplitString(
  str: string,
  outerSeparator: string,
  innerSeparator: string
): Array<ScopedAccessTokenConfiguration> {
  const splitStr = str.split(outerSeparator)
  return splitStr
    .map((s) => s.split(innerSeparator))
    .map((components) => {
      if (components.length !== 4) {
        throw new ValidationError(`Expecting four components: ${components}`, components)
      }

      const name = components[0]
      const secret = components[1]
      const identifier = createHash('sha256')
        .update(`${name}-${secret}`)
        .digest()
        .toString('base64')
      return {
        name,
        secret,
        publicKeyPEM: components[2] !== '' ? components[2] : null,
        publicKeyJWK: null as any,
        expiry: parseInt(components[3], 10),
        identifier,
      }
    })
    .map((config) => {
      if (config.publicKeyPEM === null) {
        return config
      }
      config.secret = Buffer.from(config.secret, 'base64').toString('utf8')
      config.publicKeyPEM = Buffer.from(config.publicKeyPEM, 'base64').toString('utf8')
      config.publicKeyJWK = pem2jwk(config.publicKeyPEM)
      return config
    })
}

/**
 * Configuration strings can contain application configurations by semicolon separating app tuples.
 * For example "ID1,secret1,name1;ID2,secret2,name2" if outer separator is ';' and inner separator is ','.
 * NOTE! Empty string components will be treated as null.
 */
export function clientApplicationsFromSplitString(
  str: string,
  outerSeparator: string,
  innerSeparator: string
): Array<ClientApplication> {
  const splitStr = str.split(outerSeparator)
  return splitStr
    .map((s) => s.split(innerSeparator))
    .map((components) => {
      if (components.length !== 3) {
        throw new ValidationError(`Expecting three components: ${components}`, components)
      }

      return {
        _id: components[0],
        _type: 'Application',
        secret: components[1].length > 0 ? components[1] : null,
        name: components[2],
      }
    })
}
