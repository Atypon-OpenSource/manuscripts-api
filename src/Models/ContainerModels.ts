/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ObjectTypes, Project, UserProfile } from '@manuscripts/manuscripts-json-schema'

import { ProjectRepository } from '../DataAccess/ProjectRepository/ProjectRepository'
import { ContainerService } from '../DomainServices/Container/ContainerService'

export enum ContainerRole {
  Owner = 'Owner',
  Writer = 'Writer',
  Viewer = 'Viewer',
  Editor = 'Editor',
  Annotator = 'Annotator',
}

export interface InvitedUserData {
  email: string
  name?: string
}

export interface ContainerInvitationResponse {
  containerID: string | null
  message: string
}

export enum ContainerType {
  project = 'project',
}

export type Container = Project

export type ContainerObjectType = ObjectTypes.Project

export type ContainerRepository = ProjectRepository

export interface ContainerServiceMap {
  [k: string]: ContainerService
}

export interface PatchContainerInvitation {
  role?: ContainerRole
  message?: string
  acceptedAt?: number
  invitedUserID?: string
}

export interface PatchContainerRequest {
  userProfile?: UserProfile
  role?: ContainerRole
}
