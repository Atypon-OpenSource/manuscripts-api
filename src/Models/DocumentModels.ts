/*!
 * Â© 2024 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ManuscriptDoc, Prisma } from '@prisma/client'

import { SnapshotLabel } from './SnapshotModel'

export type Doc = Record<string, any>

export type ManuscriptDocWithSnapshots = ManuscriptDoc & {
  snapshots: SnapshotLabel[]
}

export type CreateDoc = {
  manuscript_model_id: string
  project_model_id: string
  schema_version: string
  doc: Doc
}
export type StepsData = {
  steps: unknown[]
  clientIDs: number[]
  version: number
}

export type UpdateDocument = {
  doc?: Doc
  version?: number
  steps?: Prisma.JsonObject[]
  schema_version?: string
} & (
  | { doc: Doc }
  | { version: number }
  | { steps: Prisma.JsonObject[] }
  | { schema_version: string }
)
