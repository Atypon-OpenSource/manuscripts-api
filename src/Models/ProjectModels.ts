/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export type PatchProject = {
  _id: string
  title?: string
  owners?: string[]
  writers?: string[]
  viewers?: string[]
  editors?: string[]
  annotators?: string[]
  proofers?: string[]
}
export enum ProjectUserRole {
  Owner = 'Owner',
  Writer = 'Writer',
  Viewer = 'Viewer',
  Editor = 'Editor',
  Proofer = 'Proofer',
  Annotator = 'Annotator',
}
export enum ProjectPermission {
  READ,
  UPDATE,
  DELETE,
  UPDATE_ROLES,
  CREATE_MANUSCRIPT,
}
export type ArchiveOptions = {
  getAttachments?: boolean
  onlyIDs?: boolean
  includeExt: boolean | true
  types?: string[]
}
