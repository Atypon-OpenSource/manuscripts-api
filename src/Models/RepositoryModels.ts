/*!
 * Â© 2024 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DocumentExtender } from '../DataAccess/DocumentExtender'
import { EventExtender } from '../DataAccess/EventExtender'
import { ProjectExtender } from '../DataAccess/ProjectExtender'
import { Repository } from '../DataAccess/Repository'
import { SnapshotExtender } from '../DataAccess/SnapshotExtender'
import { UserExtender } from '../DataAccess/UserExtender'

export type DB = typeof Repository.prototype.DB
export type DocumentClient = typeof Repository.prototype.documentClient
export type ProjectClient = typeof Repository.prototype.projectClient
export type UserClient = typeof Repository.prototype.userClient
export type SnapshotClient = typeof Repository.prototype.snapshotClient
export type EventClient = typeof Repository.prototype.eventClient
export type DocumentExtension = ReturnType<typeof DocumentExtender.prototype.getExtension>
export type SnapshotExtension = ReturnType<typeof SnapshotExtender.prototype.getExtension>
export type UserExtension = ReturnType<typeof UserExtender.prototype.getExtension>
export type ProjectExtension = ReturnType<typeof ProjectExtender.prototype.getExtension>
export type EventExtension = ReturnType<typeof EventExtender.prototype.getExtension>

export type Extender =
  | DocumentExtender
  | SnapshotExtender
  | EventExtender
  | ProjectExtender
  | UserExtender

export type Extension =
  | DocumentExtension
  | SnapshotExtension
  | EventExtension
  | ProjectExtension
  | UserExtension

export type ExtensionMap = {
  DocumentExtender: DocumentExtension
  SnapshotExtender: SnapshotExtension
  EventExtender: EventExtension
  ProjectExtender: ProjectExtension
  UserExtender: UserExtension
}
export type GetExtension<T extends Extender> = T extends keyof ExtensionMap
  ? ExtensionMap[T]
  : never

export enum PrismaErrorCodes {
  RecordMissing = 'P2025',
}
