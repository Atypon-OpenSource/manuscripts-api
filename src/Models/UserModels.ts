/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import moment from 'moment'

import { UserActivityEventType } from './UserEventModels'
import { BucketKey } from '../Config/ConfigurationTypes'
import { ContainerRole } from './ContainerModels'
import { ValidationError } from '../Errors'

export type IdentifiableUser = { _id: string, email?: string }

export function isIdentifiableUser (user?: any): user is IdentifiableUser {
  return typeof user !== 'undefined' && typeof user._id === 'string'
}

export function ensureValidUser (user?: any): IdentifiableUser {
  if (!isIdentifiableUser(user)) {
    throw new ValidationError('User undefined or lacks id', user)
  }
  return user
}

/**
 * Represents user's reset password credentials.
 */
export interface ResetPasswordCredentials {
  /**
   * The single use reset password.
   */
  tokenId: string
  /**
   * User's new password.
   */
  newPassword: string
  /**
   * Device id.
   */
  deviceId: string
  /**
   * Application id.
   */
  appId: string
}

export interface ChangePasswordCredentials {
  /**
   * User's Id.
   */
  userId: string
  /**
   * User's current password.
   */
  currentPassword: string
  /**
   * User's new password.
   */
  newPassword: string
  /**
   * Device id.
   */
  deviceId: string
}

/**
 * Represents user's login credentials.
 */
export interface Credentials {
  /**
   * User's email.
   */
  email: string

  /**
   * User's plain password.
   */
  password?: string

  /**
   * Device id.
   */
  deviceId: string

  /**
   * Application id.
   */
  appId: string
}

export interface ServerToServerAuthCredentials {
  /**
   * Connect's user id.
   */
  connectUserID?: string

  /**
   * User's email.
   */
  email?: string

  /**
   * Device id.
   */
  deviceId: string

  /**
   * Application id.
   */
  appId: string
}

/**
 * Represents credentials for Google login access.
 */
export interface GoogleAccessCredentials {
  /**
   * User's full name.
   */
  name: string
  /**
   * User's email.
   */
  email: string
  /**
   * Device's ID.
   */
  deviceId: string
  /**
   * Application's Id.
   */
  appId: string

  /** An optional invitation to which the account creation is a response to. */
  invitationId: string | null
  /**
   * Google Access Token.
   */
  accessToken: string
  /**
   * Google Refresh Token.
   */
  refreshToken: string
}

/**
 * Represents an user entity.
 */
export interface User {
  /**
   * User's unique id.
   */
  _id: string
  /**
   * User's name.
   */
  name: string
  /**
   * User's email.
   */
  email: string
  /**
   * Connect's user id that is stored in the sub field of connect token.
   */
  connectUserID?: string
  /**
   * timestamp represents the time when the user going to be deleted.
   */
  deleteAt?: number
}

export function userForRow (userRowData: UserRow): User {
  return {
    _id: userRowData._id,
    name: userRowData.name,
    email: userRowData.email,
    connectUserID: userRowData.connectUserID,
    ...userRowData.deleteAt && { deleteAt : userRowData.deleteAt }
  }
}

/**
 * Represents an user entity and auth.
 */
export interface AuthorizedUser {
  /**
   * User's auth token.
   */
  token: string

  /**
   * User model.
   */
  user: User

  /**
   * Session cookie for Sync Gateway
   */
  syncSessions: BucketSessions
}

/**
 * Represents all possible fields for new user object.
 */
export interface UserRow {
  /**
   * User's unique id.
   */
  _id: string
  /**
   * User's name.
   */
  name: string
  /**
   * User's unique email address.
   */
  email: string
  /**
   * Connect's user id that is stored in the sub field of connect token.
   */
  connectUserID?: string
  /**
   * timestamp represents the time when the user going to be deleted.
   */
  deleteAt?: number
}
/**
 * Represents all possible fields for new user object.
 */
export interface SignupCredentials {
  /**
   * User's full name.
   */
  name: string
  /**
   * User's unique email address.
   */
  email: string
  /**
   * User's hashed password.
   */
  password?: string
  /**
   * A token signed by a trusted source
   */
  token?: string
}

/**
 * Represents all possible fields for new user object.
 */
export interface INewUser {
  /**
   * User's unique id, if not set then db will generate new one.
   */
  _id?: string

  /**
   * User's full name.
   */
  name: string | null
  /**
   * User's unique email address.
   */
  email: string
  /**
   * Connect's user id that is stored in the sub field of connect token.
   */
  connectUserID?: string
}

/**
 * Represents all possible fields for updating already exists user object.
 */
export interface IUpdateUser {
  /**
   * User's unique id, if not set then db will generate new one.
   */
  _id: string
  /**
   * The type of the document.
   */
  _type: string
  /**
   * User's full name.
   */
  name: string
  /**
   * User's unique email address.
   */
  email: string
  /**
   * User's hashed password.
   */
  password: string
  /**
   * Determine whether the user has verified his email or not.
   */
  isVerified: boolean
  /**
   * User creation date timestamp.
   */
  creationTimestamp: number
  /**
   * Connect's user id that is stored in the sub field of connect token.
   */
  connectUserID: string
  /**
   * timestamp represents the time when the user going to be deleted.
   */
  deleteAt?: number
}

/**
 * Represents all possible fields for new user object.
 */
export interface UserToken {
  /**
   * Token's unique id.
   */
  _id: string
  /**
   * User's id.
   */
  userId: string
  /**
   * True if the user has an expiry time, false otherwise.
   */
  hasExpiry: boolean
  /**
   * User's related JWT token
   */
  token: string
  /**
   * Device's unique id.
   */
  deviceId: string
  /**
   * Application's id.
   */
  appId: string

  /**
   * IAM session id.
   */
  iamSessionID?: string

  credentials?: { google: { accessToken: string , refreshToken: string } }
}

export interface UserEmail {
  /**
   * User's unique id, which represent the email.
   */
  _id: string
}

export interface InvitationToken {
  /**
   * Token's unique id.
   */
  _id: string
  /**
   * Container id.
   */
  containerID: string
  /**
   * Invited user invitation role.
   */
  permittedRole: ContainerRole
  /**
   * Container related JWT token
   */
  token: string
}

export interface UpdateInvitationToken {
  _id?: string
}

export type BucketSessions = {
  [bucketKey in BucketKey]?: string
}

export type DeviceSessions = {
  [deviceId: string]: BucketSessions
}

export interface UserStatus {
  _id: string
  password: string
  isVerified: boolean
  blockUntil: Date | null
  /**
   * Object mapping the users deviceIds to the corresponding gateway sessionId
   */
  deviceSessions: DeviceSessions
  createdAt: Date
}

export function isBlocked (userStatus: UserStatus, date: Date): boolean {
  return userStatus.blockUntil !== null && moment(date).diff(userStatus.blockUntil) <= 0 ? true : false
}

export interface UpdateUserStatus {
  deviceSessions?: DeviceSessions
  _id?: string
  password?: string
  isVerified?: boolean
  blockUntil?: Date | null
  createdAt?: Date
}

export interface UserStatusViewFunctionDocument {
  _type: string
  eventType: UserActivityEventType
  userId: string
  timestamp: number
}

export interface CreatedIAMDetails {
  user: User
  userStatus: UserStatus
}
