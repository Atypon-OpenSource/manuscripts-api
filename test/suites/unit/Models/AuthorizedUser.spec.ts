/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AuthorizedUser, userForRow } from '../../../../src/Models/UserModels'
import { authorizedUserRowData } from '../../../data/fixtures/authorizedUserRowData'
import { userRowData } from '../../../data/fixtures/userRowData'

describe('Authorized User Model', () => {
  test('should create model', () => {
    const user = userForRow(userRowData)
    const authorizedUser: AuthorizedUser = {
      token: authorizedUserRowData.token,
      user: user,
    }
    expect(authorizedUser).toMatchSnapshot()
  })
})
