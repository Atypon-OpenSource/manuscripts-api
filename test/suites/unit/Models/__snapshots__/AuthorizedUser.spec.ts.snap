// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorized User Model should create model 1`] = `
Object {
  "syncSessions": Object {
    "data": "fooo",
  },
  "token": "12345",
  "user": Object {
    "_id": "1",
    "connectUserID": undefined,
    "email": "abarmawi@live.com",
    "name": "Abdallah",
  },
}
`;
